File: tests/taskStatus.test.ts
import { describe, expect, it } from "vitest";
import { STATUS_LABELS, partitionTasks } from "@/utils/taskStatus";
import type { TaskStatus } from "@/types/task";

function row(id: string, status?: TaskStatus) {
  return { task: { id, status }, role: "owner" as const };
}

describe("partitionTasks", () => {
  it("分成進行中與已封存兩堆", () => {
    const result = partitionTasks([row("a", "active"), row("b", "archived"), row("c", "active")]);
    expect(result.active.map(item => item.task.id)).toEqual(["a", "c"]);
    expect(result.archived.map(item => item.task.id)).toEqual(["b"]);
  });

  it("已刪除的不出現在任何一堆 —— 使用者不該再看到它", () => {
    const result = partitionTasks([row("a", "active"), row("gone", "deleted")]);
    expect(result.active.map(item => item.task.id)).toEqual(["a"]);
    expect(result.archived).toEqual([]);
  });

  it("沒有 status 的舊資料當成進行中，不會憑空消失", () => {
    const result = partitionTasks([row("legacy", undefined)]);
    expect(result.active.map(item => item.task.id)).toEqual(["legacy"]);
  });

  it("空清單回傳兩個空陣列，不是 undefined", () => {
    expect(partitionTasks([])).toEqual({ active: [], archived: [] });
  });

  it("保留原本的順序", () => {
    const result = partitionTasks([row("c", "active"), row("a", "active"), row("b", "active")]);
    expect(result.active.map(item => item.task.id)).toEqual(["c", "a", "b"]);
  });

  it("原樣帶過呼叫端自己的欄位，不只回傳 task", () => {
    const result = partitionTasks([row("a", "active")]);
    expect(result.active[0].role).toBe("owner");
  });
});

describe("STATUS_LABELS", () => {
  it("三個狀態都有中文標籤 —— 畫面上不該出現英文的 active", () => {
    expect(STATUS_LABELS.active).toBe("進行中");
    expect(STATUS_LABELS.archived).toBe("已封存");
    expect(STATUS_LABELS.deleted).toBe("已刪除");
  });
});

File: flutter_app/test/task_status_test.dart
import 'package:test/test.dart';
import 'package:splitflow/domain/task_status.dart';

/// `tests/taskStatus.test.ts` 與 `taskRole.test.ts` 的 Dart 版。
void main() {
  group('taskStatusFrom', () {
    test('認得三種狀態', () {
      expect(taskStatusFrom('active'), TaskStatus.active);
      expect(taskStatusFrom('archived'), TaskStatus.archived);
      expect(taskStatusFrom('deleted'), TaskStatus.deleted);
    });

    test('沒有 status 的舊資料當成進行中 —— 不能讓它們從列表消失', () {
      expect(taskStatusFrom(null), TaskStatus.active);
      expect(taskStatusFrom(''), TaskStatus.active);
      expect(taskStatusFrom('沒看過的值'), TaskStatus.active);
    });
  });

  group('partitionTasks', () {
    // 測試用的最小形狀：只有 id 跟狀態。
    final rows = [
      ('t1', TaskStatus.active),
      ('t2', TaskStatus.archived),
      ('t3', TaskStatus.deleted),
      ('t4', TaskStatus.active),
    ];

    test('已刪除的絕對不會出現在任何一堆', () {
      final result = partitionTasks(rows, (row) => row.$2);
      final all = [...result.active, ...result.archived].map((r) => r.$1);
      expect(all, isNot(contains('t3')));
    });

    test('進行中與封存分開', () {
      final result = partitionTasks(rows, (row) => row.$2);
      expect(result.active.map((r) => r.$1).toList(), ['t1', 't4']);
      expect(result.archived.map((r) => r.$1).toList(), ['t2']);
    });

    test('保留原本的順序', () {
      final many = [
        ('a', TaskStatus.active),
        ('b', TaskStatus.active),
        ('c', TaskStatus.active),
      ];
      final result = partitionTasks(many, (row) => row.$2);
      expect(result.active.map((r) => r.$1).toList(), ['a', 'b', 'c']);
    });

    test('空清單不會爆', () {
      final result = partitionTasks(<(String, TaskStatus)>[], (row) => row.$2);
      expect(result.active, isEmpty);
      expect(result.archived, isEmpty);
    });
  });

  group('taskRole', () {
    test('owner 優先於 adminIds', () {
      // owner 通常也在 adminIds 裡，不能因此被判成 admin。
      expect(
        taskRole(ownerId: 'a', adminIds: ['a', 'b'], uid: 'a'),
        TaskRole.owner,
      );
    });

    test('在 adminIds 裡的是 admin', () {
      expect(
        taskRole(ownerId: 'a', adminIds: ['a', 'b'], uid: 'b'),
        TaskRole.admin,
      );
    });

    test('其他人是一般成員', () {
      expect(
        taskRole(ownerId: 'a', adminIds: ['a'], uid: 'c'),
        TaskRole.member,
      );
    });

    test('adminIds 是空的也算得出來', () {
      expect(taskRole(ownerId: 'a', adminIds: [], uid: 'c'), TaskRole.member);
    });
  });
}
